File: 02-Literate-Companion-Implementations/examples/ru/05-jupyter/primes.py
PRIME_COUNT = 1000
ROWS_PER_PAGE = 50
COLUMNS_PER_PAGE = 4
COLUMN_WIDTH = 10


def generate_primes(count: int) -> list[int]:
    primes = []
    if count == 0:
        return primes
    primes.append(2)
    candidate = 3
    while len(primes) < count:
        is_prime = True
        for p in primes:
            if p > candidate // p:
                break
            if candidate % p == 0:
                is_prime = False
                break
        if is_prime:
            primes.append(candidate)
        candidate += 2
    return primes


def print_table(primes: list[int]) -> None:
    page_number = 1
    page_offset = 0
    while page_offset < len(primes):
        print(f"The First {len(primes)} Prime Numbers --- Page {page_number}\n")
        for row in range(ROWS_PER_PAGE):
            line = []
            for column in range(COLUMNS_PER_PAGE):
                index = page_offset + row + column * ROWS_PER_PAGE
                if index < len(primes):
                    line.append(f"{primes[index]:>{COLUMN_WIDTH}}")
            print("".join(line))
        print()
        page_number += 1
        page_offset += ROWS_PER_PAGE * COLUMNS_PER_PAGE


if __name__ == "__main__":
    print_table(generate_primes(PRIME_COUNT))

File: 02-Literate-Companion-Implementations/examples/uk/02-noweb-like/primes.cpp
#include <cstddef>
#include <iomanip>
#include <iostream>
#include <vector>

constexpr std::size_t PRIME_COUNT = 1000;
constexpr std::size_t ROWS_PER_PAGE = 50;
constexpr std::size_t COLUMNS_PER_PAGE = 4;
constexpr int COLUMN_WIDTH = 10;
using PrimeTable = std::vector<int>;
PrimeTable generate_primes(std::size_t count);
bool is_prime_candidate(int candidate, const PrimeTable& primes);
void print_table(const PrimeTable& primes);

int main() {
    const auto primes = generate_primes(PRIME_COUNT);
    print_table(primes);
    return 0;
}

PrimeTable generate_primes(std::size_t count) {
    PrimeTable primes;
    if (count == 0) return primes;
    primes.push_back(2);

    for (int candidate = 3; primes.size() < count; candidate += 2) {
        if (is_prime_candidate(candidate, primes)) {
            primes.push_back(candidate);
        }
    }
    return primes;
}

bool is_prime_candidate(int candidate, const PrimeTable& primes) {
    for (int p : primes) {
        if (p > candidate / p) return true;
        if (candidate % p == 0) return false;
    }
    return true;
}
void print_table(const PrimeTable& primes) {
    std::size_t page_number = 1;
    std::size_t page_offset = 0;

    while (page_offset < primes.size()) {
        std::cout << "The First " << primes.size()
                  << " Prime Numbers --- Page " << page_number << "\n\n";

        for (std::size_t row = 0; row < ROWS_PER_PAGE; ++row) {
            for (std::size_t column = 0; column < COLUMNS_PER_PAGE; ++column) {
                const std::size_t index = page_offset + row + column * ROWS_PER_PAGE;
                if (index < primes.size()) {
                    std::cout << std::setw(COLUMN_WIDTH) << primes[index];
                }
            }
            std::cout << '\n';
        }

        std::cout << '\n';
        ++page_number;
        page_offset += ROWS_PER_PAGE * COLUMNS_PER_PAGE;
    }
}
